src/infra/core/src/ingest/fetch_service.rs

```diff
@@ -722,6 +729,120 @@ impl FetchService {
         }))
     }
 
+    async fn fetch_mqtt(
+        dataset_handle: &DatasetHandle,
+        fetch: &FetchStepMqtt,
+        target_path: &Path,
+        listener: &Arc<dyn FetchProgressListener>,
+    ) -> Result<FetchResult, PollingIngestError> {
+        use std::io::Write as _;
+
+        use rumqttc::{AsyncClient, Event, MqttOptions, Packet, QoS};
+
+        // TODO: Reconsider assignment of client identity on which QoS session data
+        // rentention is based
+        let client_id = format!("kamu-ingest-{}", dataset_handle.id.as_multibase());
+
+        let mut opts = MqttOptions::new(client_id, &fetch.host, u16::try_from(fetch.port).unwrap());
+        opts.set_clean_session(false);
+
+        // TODO: Reconsider password propagation
+        if let (Some(username), Some(password)) = (&fetch.username, &fetch.password) {
+            let password = Self::template_string(password)?;
+            opts.set_credentials(username, password);
+        }
+
+        tracing::debug!("Connecting to the MQTT broker and subscribing to the topic");
+
+        let (client, mut event_loop) = AsyncClient::new(opts, 1000);
+        client
+            .subscribe_many(fetch.topics.clone().into_iter().map(|s| {
+                rumqttc::SubscribeFilter::new(
+                    s.path,
+                    match s.qos {
+                        None | Some(MqttQos::AtMostOnce) => QoS::AtMostOnce,
+                        Some(MqttQos::AtLeastOnce) => QoS::AtLeastOnce,
+                        Some(MqttQos::ExactlyOnce) => QoS::ExactlyOnce,
+                    },
+                )
+            }))
+            .await
+            .int_err()?;
+
+        let mut fetched_bytes = 0;
+        let mut fetched_records = 0;
+        let mut file = std::fs::File::create(target_path).int_err()?;
+
+        // TODO: Reading from env vars is temporary - should be replaced by vars
+        // provider
+        let max_records: u64 = std::env::var(ODF_BATCH_SIZE)
```

Comment:
please add participating env variables in `.env`

Reply:
Didn't know that we have such a file - added.